third_party/move/move-compiler-v2/tests/bytecode-generator/bug_14629.move

```diff
@@ -0,0 +1,17 @@
+module 0x8675309::M {
+    const MAX_U64: u128 = 18446744073709551615;
+    struct R<T: key>  { r: T }
+    struct X<T> has key, drop {
+        r: T
+    }
+
+    fun t0() {
+        let y = R { r: X { r: 0} };
+        let R { r: _r } = y;
+    }
+
+    fun t0_u128() {
+        let y = R { r: X { r: MAX_U64 + 1} };
```

Comment:
I'm sorry I was unclear.  Inline the number here:
```
let y = R { r: X { r: 18446744073709551616 } };
```
so we can see what happens.  I would expect that the `u128` type is constraining the value in that situation.  But maybe a less direct situation is needed.

And now that I ask for this I suddenly realize I'm not sure what Move specifies it will do with a big constant.

Reply:
This will lead to the error 
```
+Diagnostics:
+error: unable to infer instantiation of type `store + u128|u256` (consider providing type arguments or annotating the type)
+   ┌─ tests/bytecode-generator/bug_14629.move:14:31
+   │
+14 │         let y = R { r: X { r: 18446744073709551616 } };
+   │                               ^^^^^^^^^^^^^^^^^^^^
```
,which I think should be fine? @brmataptos

Reply:
Another example:

```
module 0x8675309::M {
    const MAX_U64: u64 = 18446744073709551615;
    fun t0_u128() {
        let y = R { r: X { r: MAX_U64 + 1 } };
        let R { r: _r } = y;
    }
```

In general, should the compiler infer the type as u64, which will lead to overflow? I think it should be fine as well because it can only be detected during runtime?

Reply:
These are both ok.  I just wanted to make sure it didn't do something crazy like inferring `u64` while looking for abilities and then complaining about it to the user.